File: Activity_Hooks/app/src/main/java/com/example/activity_hooks/MainActivity.java
package com.example.activity_hooks;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.os.PersistableBundle;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

    public final static String EXTRA_TEXT = "com.example.activity_hooks.TEXT";

    String currentState;
    int score;
    int lives;


    public void onExplicitIntent(View view){
        Intent explicitIntent = new Intent();
        explicitIntent.setAction(Intent.ACTION_SEND);
        explicitIntent.putExtra(EXTRA_TEXT, "DEMO TEXT");
        explicitIntent.setType("text/plain");
//        MIME

        Intent window = Intent.createChooser(explicitIntent, "Chose Application");
        if (explicitIntent.resolveActivity(getPackageManager()) != null){
            startActivity(window);
        }
        else {
            Toast.makeText(this, "Fail", Toast.LENGTH_LONG).show();
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState, @NonNull PersistableBundle outPersistentState) {
        super.onSaveInstanceState(outState, outPersistentState);

        outState.putString("STATE", "SAVED");
        outState.putInt("SCORE", 4);
        outState.putInt("LIVE", 3);
    }

    @Override
    protected void onRestoreInstanceState(@NonNull Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);

        if( savedInstanceState != null ){
            currentState = savedInstanceState.getString("STATE");
            score = savedInstanceState.getInt("SCORE", 0);
            lives = savedInstanceState.getInt("LIVE", 0);
        }
        else {
//            start from scratch

        }
    }

    @Override
    protected void onStart() {
        super.onStart();
        Log.i("DESTROY", "APP :: MainActivity.onStart ");

    }

    @Override
    protected void onPause() {
        super.onPause();
        Log.i("DESTROY", "APP :: MainActivity.onPause ");
    }

    @Override
    protected void onResume() {
        super.onResume();
        Log.i("DESTROY", "APP :: MainActivity.onResume ");
    }

    @Override
    protected void onStop() {
        super.onStop();
        Log.i("DESTROY", "APP :: MainActivity.onStop");
    }

    @Override
    protected void onRestart() {
        super.onRestart();
        Log.i("DESTROY", "APP :: MainActivity.onRestart");

    }
}

File: Email_Password_Firebase_G25/app/src/main/java/com/example/email_password_firebase_g25/MainActivity.kt
package com.example.email_password_firebase_g25

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import com.google.firebase.auth.FirebaseAuth

class MainActivity : AppCompatActivity() {
    lateinit var logOutButton: Button
    lateinit var updateProfileButton: Button
    lateinit var displayNameTextView: TextView
    private lateinit var auth: FirebaseAuth


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        logOutButton = findViewById(R.id.logOutButton)
        updateProfileButton = findViewById(R.id.updateProfileButton)
        displayNameTextView = findViewById(R.id.displayNameTextView)
        auth = FirebaseAuth.getInstance()

        logOutButton.setOnClickListener {
            auth.signOut()

            val intent = Intent(this, RegisterActivity::class.java)
            startActivity(intent)
            finish()
        }

        updateProfileButton.setOnClickListener {
            val intent = Intent(this, UpdateProfileActivity::class.java)
            startActivity(intent)
        }


        if( auth.currentUser != null ){
            val user = auth.currentUser;
            displayNameTextView.text = user?.displayName
            Toast.makeText(this, "უკვე ავტორიზებული", Toast.LENGTH_LONG).show()
        }
    }
}